File: 2024/07 July/Closest Three Sum/main.py
# Link: https://www.geeksforgeeks.org/problems/three-sum-closest/1

import sys


class Solution:
    def threeSumClosest(self, arr, target):
        arr.sort()
        prev_diff = sys.maxsize
        closest_sum = -sys.maxsize - 1

        for i in range(len(arr) - 2):
            start = i + 1
            end = len(arr) - 1

            while start < end:
                sum_val = arr[i] + arr[start] + arr[end]

                if sum_val == target:
                    return target

                current_diff = abs(sum_val - target)

                if current_diff < prev_diff or (
                    current_diff == prev_diff and sum_val > closest_sum
                ):
                    prev_diff = current_diff
                    closest_sum = sum_val

                if sum_val < target:
                    start += 1
                else:
                    end -= 1

        return closest_sum


if __name__ == "__main__":
    obj = Solution()
    arr = [-7, 9, 8, 3, 1, 1]
    target = 2
    print(obj.threeSumClosest(arr, target))

    arr = [5, 2, 7, 5]
    target = 13
    print(obj.threeSumClosest(arr, target))

File: 2024/07 July/Closest Three Sum/main.cpp
// Link: https://www.geeksforgeeks.org/problems/three-sum-closest/1

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution
{
public:
    int threeSumClosest(std::vector<int> &arr, int target)
    {
        sort(arr.begin(), arr.end());

        int prevDiff = INT_MAX;
        int closestSum = INT_MIN;

        for (int i = 0; i < arr.size() - 2; i++)
        {
            int start = i + 1;
            int end = arr.size() - 1;

            while (start < end)
            {
                int sum = arr[i] + arr[start] + arr[end];

                if (sum == target)
                    return sum;

                int currentDiff = abs(sum - target);

                if (currentDiff < prevDiff || (currentDiff == prevDiff && sum > closestSum))
                {
                    prevDiff = currentDiff;
                    closestSum = sum;
                }

                if (sum < target)
                {
                    start++;
                }
                else
                {
                    end--;
                }
            }
        }
        return closestSum;
    }
};

int main()
{
    Solution obj;
    vector<int> arr = {-7, 9, 8, 3, 1, 1};
    int target = 2;
    cout << obj.threeSumClosest(arr, target) << endl;

    arr = {5, 2, 7, 5};
    target = 13;
    cout << obj.threeSumClosest(arr, target) << endl;

    return 0;
}